ddsrouter_utils/src/cpp/thread_pool/pool/SlotThreadPool.cpp

```diff
@@ -126,22 +126,24 @@ void SlotThreadPool::thread_routine_()
             logDebug(DDSROUTER_THREAD_POOL, "Thread: " << std::this_thread::get_id() << " free, getting new callback.");
             TaskId task_id = task_queue_.consume();
 
-            // Lock to access the slot map
-            slots_mutex_.lock();
-
-            auto it = slots_.find(task_id);
-            // Check the slot is correct
-            if (it == slots_.end())
             {
-                utils::tsnh(STR_ENTRY << "Slot in Queue must be stored in slots register");
+                // Lock to access the slot map while executing task
+                // NOTE: this can end before calling task if the task object is copied, but if it is used with
+                // a reference, it cannot unlock before calling method
+                std::shared_lock<SlotsMapType> lock(slots_);
+
+                auto it = slots_.find(task_id);
+                // Check the slot is correct
+                if (it == slots_.end())
+                {
+                    throw utils::ValueNotAllowedException(STR_ENTRY << "Slot " << task_id << " in queue not registered.");
+                }
+
+                Task& task = it->second;
+
+                logDebug(DDSROUTER_THREAD_POOL, "Thread: " << std::this_thread::get_id() << " executing callback.");
+                task();
```

Comment:
I would say that the time taken for copying a function (in our case, only with functions whose args are not copied) is not so relevant. However, it might be problematic to have the discovery thread stuck for not being able to slot a new task.